### PR TITLE
feat(server): implement limited scope and token management for public API

### PR DESCRIPTION
**Pull request by mimoham24**
## Release Notes

- **New Features**
  - Introduced a new mutation for regenerating public API tokens.
  - Added support for optional tokens in project publications.
  
- **Improvements**
  - Enhanced GraphQL schema to reflect new token functionalities.
  - Updated error handling for project publication scopes.
  - Improved authentication middleware for public API access.
  - Streamlined token generation logic by utilizing a standard library.

- **Bug Fixes**
  - Resolved issues with token retrieval and project visibility checks.

- **Tests**
  - Added comprehensive tests for new functionalities, including token regeneration and project updates.
  - Enhanced test cases for public API authorization and project publication settings.

### CODE REVIEW

File: server/pkg/token/token.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package token
```
Comment (rot1024) [inline server/pkg/token/token.go:1]: Not enough to split the implementation into independent packages?

File: server/schemas/project.graphql [hunk 2]
```diff
@@ -12,6 +12,7 @@ enum ProjectPublicationScope {
 type ProjectPublication {
   scope: ProjectPublicationScope!
   assetPublic: Boolean!
+  token: String
```
Comment (yk-eukarya) [inline server/schemas/project.graphql:15]:
```suggestion
```
For security reasons, it's better to reveal the token **only** once it is generated. However, we need to confirm this with Nova first.

File: server/pkg/project/publication.go [hunk 3]
```diff
@@ -31,10 +42,25 @@ func (p *Publication) AssetPublic() bool {
 	return p.assetPublic
 }
 
+func (p *Publication) Token() string {
+	return p.token
+}
+
+func (p *Publication) SetToken(t string) {
+	p.token = t
+}
```
[the remark this answers is not in the document]
Comment (yk-eukarya) [reply server/pkg/project/publication.go:55]:
```suggestion
```

File: server/internal/usecase/operator.go [hunk 4]
```diff
@@ -19,6 +19,7 @@ type Operator struct {
 	WritableProjects     project.IDList
 	OwningProjects       project.IDList
 	MaintainableProjects project.IDList
+	PublicAPIToken       string
```
Comment (yk-eukarya) [inline server/internal/usecase/operator.go:22]:
```suggestion
```
the operator does not contain any token, but the result of the auth stage

File: server/internal/usecase/interactor/project.go [hunk 5]
```diff
@@ -164,3 +165,30 @@ func (i *Project) Delete(ctx context.Context, projectID id.ProjectID, operator *
...
+			if err != nil {
+				return nil, err
+			}
+
+			// check if the user is the owner of the project
+			if !operator.IsOwningProject(p.ID()) {
+				return nil, interfaces.ErrOperationDenied
+			}
+
+			p.Publication().GenerateToken()
```
[the remark this answers is not in the document]
Comment (yk-eukarya) [reply server/internal/usecase/interactor/project.go:185]: whats happens if `GenerateToken` called on a public project?

File: server/internal/infrastructure/mongo/project.go [hunk 6]
```diff
@@ -98,6 +98,15 @@ func (r *ProjectRepo) FindByPublicName(ctx context.Context, name string) (*proje
 	})
 }
 
+func (r *ProjectRepo) FindByPublicAPIToken(ctx context.Context, token string) (*project.Project, error) {
+	if token == "" {
+		return nil, rerror.ErrNotFound
+	}
+	return r.findOne(ctx, bson.M{
+		"publication.token": token,
```
Comment (yk-eukarya) [inline server/internal/infrastructure/mongo/project.go:106]: consider updating Mongo DB collection indexes

File: server/internal/app/auth_client.go [hunk 7]
```diff
@@ -293,6 +293,34 @@ func generateIntegrationOperator(ctx context.Context, cfg *ServerConfig, i *inte
...
+		AcOperator: &accountusecase.Operator{
+			User: nil,
+			ReadableWorkspaces:     id.WorkspaceIDList{w.ID()},
+			WritableWorkspaces:     nil,
+			MaintainableWorkspaces: nil,
+			OwningWorkspaces:       nil,
+		},
+		Integration:          nil,
+		Lang:                 lang,
+		PublicAPIToken:       token,
```
Comment (yk-eukarya) [inline server/internal/app/auth_client.go:307]:
```suggestion
```

File: server/internal/app/auth_client.go [hunk 8]
```diff
@@ -293,6 +293,34 @@ func generateIntegrationOperator(ctx context.Context, cfg *ServerConfig, i *inte
...
+
+	w, err := cfg.Repos.Workspace.FindByID(ctx, p.Workspace())
+	if err != nil {
+		return nil, err
+	}
+
+	return &usecase.Operator{
+		AcOperator: &accountusecase.Operator{
+			User: nil,
+			ReadableWorkspaces:     id.WorkspaceIDList{w.ID()},
```
Comment (yk-eukarya) [inline server/internal/app/auth_client.go]: is it possible to leak another project data?

File: server/internal/adapter/publicapi/controller.go [hunk 9]
```diff
@@ -37,17 +38,16 @@ func (c *Controller) checkProject(ctx context.Context, prj string) (*project.Pro
...
-	// 	if op := adapter.Operator(ctx); op == nil || t == "" || op.PublicAPIToken != t {
-	// 		return nil, ErrInvalidProject
-	// 	}
-	// }
+	if pr.Publication().Scope() == project.PublicationScopeLimited {
+		t := pr.Publication().Token()
+		if op := adapter.Operator(ctx); op == nil || t == "" || op.PublicAPIToken != t {
+			return nil, ErrInvalidProject
+		}
+	}
```
Comment (yk-eukarya) [inline server/internal/adapter/publicapi/controller.go:50]: The check should done based on the readable projects

File: server/e2e/publicapi_test.go [hunk 10]
```diff
@@ -370,11 +371,97 @@ func TestPublicAPI(t *testing.T) {
...
+
+	// invalid token
+	e.GET("/api/p/{project}/{model}", publicAPIProjectAlias, publicAPIModelKey).
+		WithHeader("Origin", "https://example.com").
+		WithHeader("Authorization", "secret_abc").
+		WithHeader("Content-Type", "application/json").
+		Expect().
+		Status(http.StatusUnauthorized)
+
+	// valid token
```
Comment (yk-eukarya) [inline server/e2e/publicapi_test.go:393]: try to access another project within the same workspace, it should be unauthorized

File: server/internal/app/auth_client.go [hunk 11]
```diff
@@ -126,37 +127,32 @@ func attachIntegrationOperator(ctx context.Context, req *http.Request, cfg *Serv
...
-			// 	}
-
-			// 	c.SetRequest(req.WithContext(adapter.AttachOperator(ctx, &usecase.Operator{
-			// 		PublicAPIProject: ws.ID(),
-			// 	})))
-			// }
+			req := c.Request()
+			ctx := req.Context()
+			if token := getToken(req); token != "" {
+				p, err := cfg.Repos.Project.FindByPublicAPIToken(ctx, token)
```
Comment (yk-eukarya) [inline server/internal/app/auth_client.go:136]: we should discuss the project publication status here, if the project was set to private the auth should fail even if the token is valid

File: server/pkg/project/publication.go [hunk 12]
```diff
@@ -31,10 +46,25 @@ func (p *Publication) AssetPublic() bool {
...
```
Comment (yk-eukarya) [inline server/pkg/project/publication.go:55]:
```suggestion
```

File: server/internal/usecase/interactor/project.go (hunk 5 above)
Comment (yk-eukarya) [inline server/internal/usecase/interactor/project.go:185]: consider null check and different project publication status

File: server/internal/infrastructure/mongo/project.go [hunk 13]
```diff
@@ -16,7 +16,7 @@ import (
 )
 
 var (
-	projectIndexes       = []string{"workspace"}
+	projectIndexes       = []string{"workspace", "publication.token"}
```
[the remark this answers is not in the document]
Comment (yk-eukarya) [reply server/internal/infrastructure/mongo/project.go:19]: @nourbalaha handle this one plz

File: server/internal/app/auth_client.go [hunk 14]
```diff
@@ -126,37 +127,36 @@ func attachIntegrationOperator(ctx context.Context, req *http.Request, cfg *Serv
...
+				if err != nil {
+					if errors.Is(err, rerror.ErrNotFound) {
+						return c.JSON(http.StatusUnauthorized, map[string]string{"error": "invalid token"})
+					}
+					return err
+				}
+
+				if p.Publication().Scope() == project.PublicationScopePrivate {
```
Comment (yk-eukarya) [inline server/internal/app/auth_client.go:144]: possible null exception 
```suggestion
				if p.Publication() == nil || p.Publication().Scope() == project.PublicationScopePrivate {
```

File: server/internal/infrastructure/mongo/project.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nourbalaha) [reply server/internal/infrastructure/mongo/project.go:19]: When I did this I got the error below, that is why I reverted:
`2024-11-06T18:25:00.193+0900    FATAL   Failed to init mongo: (DuplicateKey) Index build failed: b5c96961-5e47-4824-9f71-d77776065a8b: Collection reearth_cms.project ( 4d7b62d4-95de-4c43-b4ce-f1fce1bfca7b ) :: caused by :: E11000 duplicate key error collection: reearth_cms.project index: publication.token_1 dup key: { publication.token: null }`

[the remark this answers is not in the document]
Comment (yk-eukarya) [reply server/internal/infrastructure/mongo/project.go:19]: The projects without a token are considered duplicated! The token is randomly generated so we can skip this one for now, but please leave a TODO comment on it so we can fix it later.

**Review by yk-eukarya — APPROVED** [review]

File: server/internal/infrastructure/mongo/project.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (nourbalaha) [reply server/internal/infrastructure/mongo/project.go:20]: TODO comment added!

**Review by nourbalaha — APPROVED** [review]

**Review by yk-eukarya — APPROVED** [review]

**Pull request merged**